src/shims/native_lib/mod.rs

```diff
@@ -267,6 +269,157 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
 
         interp_ok(())
     }
+
+    /// Extract the value from the result of reading an operand from the machine
+    /// and convert it to a `OwnedArg`.
+    fn op_to_ffi_arg(&self, v: &OpTy<'tcx>, tracing: bool) -> InterpResult<'tcx, OwnedArg> {
+        let this = self.eval_context_ref();
+        let scalar = |v| interp_ok(this.read_immediate(v)?.to_scalar());
+        interp_ok(match v.layout.ty.kind() {
+            // If the primitive provided can be converted to a type matching the type pattern
+            // then create a `OwnedArg` of this primitive value with the corresponding `OwnedArg` constructor.
+            // the ints
+            ty::Int(IntTy::I8) => ScalarArg::Int8(scalar(v)?.to_i8()?).into(),
+            ty::Int(IntTy::I16) => ScalarArg::Int16(scalar(v)?.to_i16()?).into(),
+            ty::Int(IntTy::I32) => ScalarArg::Int32(scalar(v)?.to_i32()?).into(),
+            ty::Int(IntTy::I64) => ScalarArg::Int64(scalar(v)?.to_i64()?).into(),
+            ty::Int(IntTy::Isize) =>
+                ScalarArg::ISize(scalar(v)?.to_target_isize(this)?.try_into().unwrap()).into(),
+            // the uints
+            ty::Uint(UintTy::U8) => ScalarArg::UInt8(scalar(v)?.to_u8()?).into(),
+            ty::Uint(UintTy::U16) => ScalarArg::UInt16(scalar(v)?.to_u16()?).into(),
+            ty::Uint(UintTy::U32) => ScalarArg::UInt32(scalar(v)?.to_u32()?).into(),
+            ty::Uint(UintTy::U64) => ScalarArg::UInt64(scalar(v)?.to_u64()?).into(),
+            ty::Uint(UintTy::Usize) =>
+                ScalarArg::USize(scalar(v)?.to_target_usize(this)?.try_into().unwrap()).into(),
+            ty::RawPtr(..) => {
+                let ptr = scalar(v)?.to_pointer(this)?;
+                this.expose_and_warn(ptr.provenance, tracing)?;
+
+                // This relies on the `expose_provenance` in the `visit_reachable_allocs` callback
+                // below to expose the actual interpreter-level allocation.
+                ScalarArg::RawPtr(std::ptr::with_exposed_provenance_mut(ptr.addr().bytes_usize()))
+                    .into()
+            }
+            // For ADTs, create an FfiType from their fields.
+            ty::Adt(adt_def, args) => {
+                let adt = this.adt_to_ffitype(v.layout.ty, *adt_def, args)?;
+
+                // Copy the raw bytes backing this arg.
+                let bytes = match v.as_mplace_or_imm() {
+                    either::Either::Left(mplace) => {
+                        // Get the alloc id corresponding to this mplace, alongside
+                        // a pointer that's offset to point to this particular
+                        // mplace (not one at the base addr of the allocation).
+                        let mplace_ptr = mplace.ptr();
+                        let sz = mplace.layout.size.bytes_usize();
+                        let id = this
+                            .alloc_id_from_addr(
+                                mplace_ptr.addr().bytes(),
```

Comment:
Here you are doing the equivalent of casting a pointer to an integer and back. Please don't. You can get the alloc ID from the pointer directly.

Reply:
Specifically, please use `ptr_get_alloc_id`. That also gives you the offset inside the allocation.

However, you'll need to first separately handle the zero-sized case. `mplace.ptr()` might not point to an `Allocation` if the type has size 0.

Reply:
Are ZSTs allowed over FFI? I was under the impression that there's no way to represent them in C so I figured just erroring here makes sense